regex/src/main/scala/regex/Regex.scala

```diff
@@ -8,20 +8,73 @@ package regex
 
 // Add your definitions here
 
+trait RegularLanguage
+
+case object Empty extends RegularLanguage
+
+case object Epsilon extends RegularLanguage
+
+case class Character(obj: Char) extends RegularLanguage
+
+case class Union(objOne: RegularLanguage, objTwo: RegularLanguage) extends RegularLanguage
+
+case class Concat(objOne: RegularLanguage, objTwo: RegularLanguage) extends RegularLanguage
+
+case class Star(objOne: RegularLanguage) extends RegularLanguage
+
 /** *****************************************************************************
   * Derivatives
   *
   * Fill in the function definitions below
   */
 
 /** Simplifies a regular language */
-def simplify(lang: RegularLanguage): RegularLanguage = ???
+def simplify(lang: RegularLanguage): RegularLanguage = 
+  lang match
+    case Concat(Empty, x) => simplify(Star(Empty))
+    case Concat(x, Empty) => simplify(Star(Empty))
+    case Concat(Epsilon, x) => simplify(x)
+    case Concat(x, Epsilon) => simplify(x)
+    case Concat(x, y) => Concat(simplify(x), simplify(y))
+    case Union(Empty, x) => simplify(x)
+    case Union(x, Empty) => simplify(x)
+    case Union(x, y) => Union(simplify(x), simplify(y))
+    case Star(Empty) => Empty
+    case Star(Epsilon) => Epsilon
+    case Character(x) => Character(x)
+    case Star(x) => Star(simplify(x))
 
 /** A language is nullable if it contains ε */
-def nullable(lang: RegularLanguage): Boolean = ???
+def nullable(lang: RegularLanguage): Boolean = 
+  lang match 
+    case Concat(Epsilon, Epsilon) => true
+    case Concat(Epsilon, objTwo) => false // I dont know why these aren't nullable
+    case Concat(objOne, Epsilon) => false // I dont know why these aren't nullable
+    case Concat(objOne, objTwo) => false
+    case Star(Epsilon) => true
+    case Star(objOne) => true
+    case Union(Epsilon, objTwo) => true
+    case Union(objOne, Epsilon) => true
+    case Union(objOne, objTwo) => false
+    case Epsilon => true
+    case Empty => false
+    case Character(obj) => false
 
 /** Computes the derivative of a language, with respect to a character */
-def derivative(l: RegularLanguage)(c: Char): RegularLanguage = ???
+def derivative(l: RegularLanguage)(c: Char): RegularLanguage = 
+  l match
+    case Union(x, y) => Union(derivative(x)(c), derivative(y)(c))
+    case Concat(x, y) => 
+      if nullable(x) then Union(Concat(derivative(x)(c), y), y)
+      else Concat(derivative(x)(c), y)
+    case Star(x) => Concat(derivative(x)(c), Star(x))
+    case Empty => Empty
+    case Epsilon => Empty
+    case x =>
+      if Character(c) == x then Epsilon
+      else Empty
+
```

Comment:
Nice! All three of these match our implementation as well. I think bringing the If - elses on a new line should be avoided if possible. I just personally find it harder to read